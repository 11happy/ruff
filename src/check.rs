use std::path::Path;

use rustpython_parser::ast::{Located, StmtKind, Suite};

use crate::message::Message;

pub fn check_ast(path: &Path, python_ast: &Suite) -> Vec<Message> {
    let mut messages: Vec<Message> = vec![];
    for statement in python_ast {
        let Located {
            location,
            custom: _,
            node,
        } = statement;
        match node {
            StmtKind::FunctionDef { .. } => {}
            StmtKind::AsyncFunctionDef { .. } => {}
            StmtKind::ClassDef { .. } => {}
            StmtKind::Return { .. } => {}
            StmtKind::Delete { .. } => {}
            StmtKind::Assign { .. } => {}
            StmtKind::AugAssign { .. } => {}
            StmtKind::AnnAssign { .. } => {}
            StmtKind::For { .. } => {}
            StmtKind::AsyncFor { .. } => {}
            StmtKind::While { .. } => {}
            StmtKind::If { .. } => {}
            StmtKind::With { .. } => {}
            StmtKind::AsyncWith { .. } => {}
            StmtKind::Raise { .. } => {}
            StmtKind::Try { .. } => {}
            StmtKind::Assert { .. } => {}
            StmtKind::Import { .. } => {}
            StmtKind::ImportFrom {
                level: _,
                module: _,
                names,
            } => {
                for alias in names {
                    if alias.name == "*" {
                        messages.push(Message::ImportStarUsage {
                            filename: path.to_path_buf(),
                            location: *location,
                        });
                    }
                }
            }
            StmtKind::Global { .. } => {}
            StmtKind::Nonlocal { .. } => {}
            StmtKind::Expr { .. } => {}
            StmtKind::Pass => {}
            StmtKind::Break => {}
            StmtKind::Continue => {}
        }
    }
    messages
}
