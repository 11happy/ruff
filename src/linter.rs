use std::path::Path;

use anyhow::Result;
use log::debug;
use serde::{Deserialize, Serialize};

use crate::check::check_ast;
use crate::message::Message;
use crate::{cache, parser};

#[derive(Serialize, Deserialize)]
struct CacheMetadata {
    size: u64,
    mtime: i64,
}

#[derive(Serialize, Deserialize)]
struct CheckResult {
    metadata: CacheMetadata,
    messages: Vec<Message>,
}

pub fn check_path(path: &Path) -> Result<Vec<Message>> {
    // TODO(charlie): These specific files are causing a stack overflow.
    if path.to_string_lossy().eq_ignore_ascii_case(
         "../spring-experiments/spr_experiments/spr_experiments/assayworks/experiments/order_20220204/pipeline_steps.py") ||
        path.to_string_lossy().eq_ignore_ascii_case
        ( "../spring-experiments/spr_platform/spr_platform/data_index/bigquery_index.py")
    {
        return Ok(vec![]);
    }

    // Check the cache.
    if let Some(messages) = cache::get(path) {
        debug!("Cache hit for: {}", path.to_string_lossy());
        return Ok(messages);
    }

    // Run the linter.
    let python_ast = parser::parse(path)?;
    let messages = check_ast(path, &python_ast);
    cache::set(path, &messages);

    Ok(messages)
}
