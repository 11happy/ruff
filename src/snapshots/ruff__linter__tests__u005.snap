---
source: src/linter.rs
expression: checks
---
- kind: NoAssertEquals
  location:
    row: 6
    column: 9
  end_location:
    row: 6
    column: 26
  fix:
    content: self.assertEqual
    location:
      row: 6
      column: 9
    end_location:
      row: 6
      column: 26
    applied: false
- kind: NoAssertEquals
  location:
    row: 7
    column: 9
  end_location:
    row: 7
    column: 26
  fix:
    content: self.assertEqual
    location:
      row: 7
      column: 9
    end_location:
      row: 7
      column: 26
    applied: false

